source/common/signal/fatal_error_handler.cc

```diff
@@ -66,8 +91,95 @@ void callFatalErrorHandlers(std::ostream& os) {
     for (const auto* handler : *list) {
       handler->onFatalError(os);
     }
-    delete list;
+    fatal_error_handlers.store(list, std::memory_order_release);
+  }
+}
+
+void registerFatalActions(FatalAction::FatalActionPtrList safe_actions,
+                          FatalAction::FatalActionPtrList unsafe_actions,
+                          Thread::ThreadFactory& thread_factory) {
+  // Create a FatalActionManager and try to store it. If we fail to store
+  // our manager, it'll be deleted due to the unique_ptr.
+  auto mananger = std::make_unique<FatalAction::FatalActionManager>(
+      std::move(safe_actions), std::move(unsafe_actions), thread_factory);
+  FatalAction::FatalActionManager* unset_manager = nullptr;
+
+  if (fatal_action_manager.compare_exchange_strong(unset_manager, mananger.get(),
+                                                   std::memory_order_acq_rel)) {
+    // Our manager is the system's singleton, ensure that the unique_ptr does not
+    // delete the instance.
+    mananger.release();
+  } else {
+    ENVOY_BUG(false, "Fatal Actions have already been registered.");
+  }
+}
+
+FatalAction::Status runSafeActions() {
+  // Check that registerFatalActions has already been called.
+  FatalAction::FatalActionManager* action_manager =
+      fatal_action_manager.load(std::memory_order_acquire);
+
+  if (action_manager == nullptr) {
+    return FatalAction::Status::ActionManangerUnset;
+  }
+
+  // Check that we're the thread that gets to run the actions.
+  int64_t my_tid = action_manager->getThreadFactory().currentThreadId().getId();
+  int64_t expected_tid = -1;
+
+  if (failure_tid.compare_exchange_strong(expected_tid, my_tid, std::memory_order_acq_rel,
+                                          std::memory_order_acquire)) {
+    // Run the actions
+    runFatalActions(action_manager->getSafeActions());
+    return FatalAction::Status::Success;
+  } else if (expected_tid == my_tid) {
+    return FatalAction::Status::AlreadyRanOnThisThread;
+  }
+
+  return FatalAction::Status::RunningOnAnotherThread;
+}
+
+FatalAction::Status runUnsafeActions() {
```

Comment:
This is basically the same function as the one above. Can you have a functor or pass whether to run safe or unsafe to a common function?